Pattern Validation - Max Length Issue Fix

samyakOO7 opened the pull request:
Pattern validation was taking priority in response schema for stub and request schema for test over max length parameter due to random() method generating value more than default

Fixed the use case putting a constraint of max length on random() method so that it never generates value for than max length keeping the current logic intact


Test Spec File - [validate-regex-spec.yaml.zip](https://github.com/user-attachments/files/17747959/validate-regex-spec.yaml.zip)

Backward Compatibility Spec (v2.0.36 release check) - 
[validate-regex-spec.yaml.zip](https://github.com/user-attachments/files/17797468/validate-regex-spec.yaml.zip)


Logic used -

>Default length - 5 (remains intact)

> If minimum and maximum is both supplied, pattern length would be in range of (min to max) 

> If minimum length is not supplied in spec -> and maximum is less than default (5), pattern would be of length (1 i.e. generex min value, max)
Say max = 4 is only supplied 
pattern length would be in range (1-4)

> If minimum length is not supplied in spec -> and maximum is greater than default (5), pattern would be of length (5 (default, max)
Say max = 15 is only supplied 
pattern length would be in range (5-15)

> if max is not supplied and min value exist we will assume to take pattern length from (min , generex max value)

**Test for maximum length conditions :**
Request : `{minLength : 2, maxLength: 11}`
Response : `{minLength : 2, maxLength: 6}`


Before Stub : 
```shell
curl -X POST -H "Content-Type: application/json" -d '{ "number": "12000000" }' http://0.0.0.0:49272/validate
{
    "number": "873.13863"
}
```               
Generated value : 9 digit

Before Test:
Details:
```shell
Scenario: POST /validate -> 200 has FAILED
In scenario "Validate the input number. Response: Input is valid"
API: POST /validate -> 200

  In scenario "Validate the input number. Response: Input is valid"
     API: POST /validate -> 200
   
       REQUEST.BODY.number Contract expected string with maxLength 2 but request contained "2.8837947"
```
          
After Issue Fixed Stub:

```shell
curl -X POST -H "Content-Type: application/json" -d '{ "number": "12000000" }' http://0.0.0.0:9000/validate
{
    "number": "8.9162"
}
```           

After Issue Fixed Test:

```shell
Request:
POST /validate
Accept-Charset: UTF-8
Accept: */*
Content-Type: application/json

{
    "number": "3."
}

Response:
200 OK
Vary: Origin
X-Specmatic-Result: success
X-Specmatic-Type: random
Content-Length: 26
Content-Type: application/json
Connection: keep-alive

{
    "number": "3618.9"
}
```

**Test for Minimum Length  conditions:**
Request : `{minLength : 2, maxLength: 11}`
Response : `{minLength : 2, maxLength: 6}`

Stub:
Before Fix (The output generated was defaulting to 5 as minimum for num<5 (breaching min validation):
```shell
curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:54811/validate
{
    "number": "2.866"
}
 curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:54811/validate
{
    "number": "235.4"
}
curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:54811/validate
{
    "number": "8.150"
}
```   


After Fix (Min validation is respected):  
```shell
curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:9000/validate
{
    "number": "2."
}
 curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:9000/validate
{
    "number": "5.1"
}
curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:54811/validate
{
    "number": "8.150"
}
 curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:9000/validate
{
    "number": "39"
}
```   


Test :
Before Fix: 
```shell
Request:
POST /validate
Accept-Charset: UTF-8
Accept: */*
Content-Type: application/json

{
    "number": "6.7293812038"
} 

Response:
400 Bad Request
Vary: Origin
X-Specmatic-Result: failure
Content-Length: 411
Content-Type: text/plain
Connection: keep-alive

In scenario "Validate the input number. Response: Input is valid"
API: POST /validate -> 200

  >> REQUEST.BODY.number
  
     Contract expected string with maxLength 2 but request contained "6.7293812038"
```
After Fix : 
```shell
Request:
POST /validate
Accept-Charset: UTF-8
Accept: */*
Content-Type: application/json

{
    "number": "9."
}
Response:
200 OK
Vary: Origin
X-Specmatic-Result: success
X-Specmatic-Type: random
Content-Length: 23
Content-Type: application/json
Connection: keep-alive

{
    "number": "407"
}
```
Some edge cases taken care of defaults -

 Request {  maxLength = 4 }
Response {  maxLength = 4 }  
min  = null and max (4) < default (5) [pattern (min ‎ =  Generex(min val = 1), max = 4)]

Stub- 
```shell
curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:9000/validate
{
    "number": "7"
} curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:9000/validate
{
    "number": "3.47"
}
```   
Test - 
```shell
Request:
POST /validate
Accept-Charset: UTF-8
Accept: */*
Content-Type: application/json

{
    "number": "65.0"
}
Response:
200 OK
Vary: Origin
X-Specmatic-Result: success
X-Specmatic-Type: random
Content-Length: 23
Content-Type: application/json
Connection: keep-alive

{
    "number": "4.72”
}
```
Request {  maxLength = 15 }
Response {  maxLength = 15 }
min length = null and max length > default (5) [pattern (min ‎ =  default(5), max = 15)]

Stub-
```shell
curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:9000/validate
{
    "number": "8.915"
} curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:9000/validate
{
    "number": "9.609"
}
curl -X POST -H "Content-Type: application/json" -d '{ "number": "1" }' http://0.0.0.0:9000/validate
{
    "number": "7.09024600"
}
```   

Test - 
``` shell
Request:
POST /validate
Accept-Charset: UTF-8
Accept: */*
Content-Type: application/json

{
    "number": "4.22214"
} Response:
200 OK
Vary: Origin
X-Specmatic-Result: success
X-Specmatic-Type: random
Content-Length: 25
Content-Type: application/json
Connection: keep-alive

{
    "number": "2.648"
} 
```
Clear outputs respecting validation schema are found with multiple test of min and max length

Thanks

Commit:
Pattern Validation - Max Length Issue Fix

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/core/pattern/StringPattern.kt:
```diff
@@ -67,20 +67,28 @@ data class StringPattern (
-            if(defaultExample == null)
-                return StringValue(Generex(regex.removePrefix("^").removeSuffix("$")).random(randomStringLength))
+            if(defaultExample == null) {
+                if (maxLength != null)
```
Is there a possibility that the regex pattern can generate a String that is less than `minLength`, should we handle that also?

harikrishnan83 commented on core/src/test/kotlin/io/specmatic/core/pattern/StringPatternTest.kt:
```diff
@@ -240,4 +250,12 @@ internal class StringPatternTest {
     }
+
+    @Test
+    fun `should fail to generate string when maxLength is less than minLength`() {
```
good that you have covered this scenario with a test

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/core/pattern/StringPattern.kt:
```diff
@@ -67,20 +67,28 @@ data class StringPattern (
+            if(defaultExample == null) {
+                if (maxLength != null)
+                    return StringValue(
+                        Generex(regex.removePrefix("^").removeSuffix("$")).random(
```
any reason we have to remove the `^` and `$`?

samyakOO7 replied:
As these characters are not addressed by generex we are omitting it

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/core/pattern/StringPattern.kt:
```diff
@@ -67,20 +67,28 @@ data class StringPattern (
+                if (maxLength != null)
+                    return StringValue(
+                        Generex(regex.removePrefix("^").removeSuffix("$")).random(
+                            randomStringLength,
```
if `randomStringLength` is already based on `minLength` and `maxLength` why are we looking at `maxLength` again here?

samyakOO7 replied:
Before it was only based on minLength

Commit:
Min Length Fix Tested

Commit:
Default case updated

Commit:
Test case fixed

Commit:
Cleaned code- Removed Cognitive Complexity

Commit:
Pattern Validation Logic for Minimum Size Regex

Commit:
Pattern Validation Logic for Minimum Size Regex

Commit:
Updated Default Example Match Condition To the Top

harikrishnan83 committed:
#1417 making test and stub consistent when pattern is defined for String
- Removing comments and instead renaming variables for readability
- Removing redundant rows in parameterised test for min and max length based String generation
- Adding test for regex based generation
- Extracting regex generation code to separate method

harikrishnan83 reviewed (approved):
Looks good to me

Commit:
File formatted

Commit:
File formatted

Commit:
File formatted

Commit:
Fixed test cases

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/StringPattern.kt:
```diff
@@ -67,36 +67,27 @@ data class StringPattern (
+            if (matches(it, resolver).isSuccess()) {
+                return it
+            }
+            throw ContractException("Schema example ${it.toStringLiteral()} does not match pattern $regex")
```
In this form, this condition is no more about regex matching.

Ideally, the resolver should be given a new MismatchMessages indicating that we are checking the schema example.

The `matches` method returns a result, and we could call result.throwOnFailure() instead of throwing this exception

samyakOO7 replied:
Resolved

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/StringPattern.kt:
```diff
@@ -67,36 +67,27 @@ data class StringPattern (
-            minLength != null && 5 < minLength -> minLength
-            maxLength != null && 5 > maxLength -> maxLength
+            minLength != null && minLength > 0 -> minLength
+            maxLength != null && maxLength < 5 -> 1
```
This worked a little differently earlier. If maxLength < 4, use it, else use upto 5 chars. What's the reason we changed this?

samyakOO7 replied:
Logic used -

Default length - 5 (remains intact)
If minimum and maximum is both supplied, pattern length would be in range of (min to max)
If minimum length is not supplied in spec -> and maximum is less than default (5), pattern would be of length (1 i.e. generex min value, max)
Say max = 4 is only supplied
pattern length would be in range (1-4)
If minimum length is not supplied in spec -> and maximum is greater than default (5), pattern would be of length (5 (default, max)
Say max = 15 is only supplied
pattern length would be in range (5-15)
if max is not supplied and min value exist we will assume to take pattern length from (min , generex max value)

samyakOO7 replied:
We revamped this logic a bit as per the one given above

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/StringPattern.kt:
```diff
@@ -157,6 +148,11 @@ data class StringPattern (
+
+    private fun generateFromRegex(regexWithoutCaretAndDollar: String, minLength: Int, maxLength: Int?): String =
+        maxLength?.let {
+            Generex(regexWithoutCaretAndDollar).random(minLength, it)
```
Regexes supported by Generex may also have a length, e.g. a{10}. While we are at it, do we want to handle this in case it's greater than a maxLength or less than minLength?

samyakOO7 replied:
We are putting the constraints to generate regex from min to max length, previously we only gave it the min length and we assumed it to be normal length, which caused it to overflow most of the time, out of max range

The logic written before was a bit tricky to understand but was failing lot of cases

joelrosario commented on core/src/test/kotlin/io/specmatic/core/pattern/StringPatternTest.kt:
```diff
@@ -72,28 +72,47 @@ internal class StringPatternTest {
-    @MethodSource("lengthTestValues")
-    fun `generate string value of appropriate length matching minLength and maxLength parameters`(min: Int?, max: Int?, length: Int) {
+    @MethodSource("minLengthMaxLengthAndExpectedLength")
+    fun `generate string value as per minLength and maxLength`(min: Int?, max: Int?, expectedLengthOfGeneratedValue: Int) {
```
Use CsvSource instead of MethodSource, to keep the test data local to the test.

joelrosario commented on core/src/test/kotlin/io/specmatic/core/pattern/StringPatternTest.kt:
```diff
@@ -72,28 +72,47 @@ internal class StringPatternTest {
+
+    @ParameterizedTest
+    @MethodSource("regexMinLengthAndMaxLengthAndExpectedLength")
+    fun `generate string value as per regex in conjunction with minLength and maxLength`(regex: String?, min: Int?, max: Int?, expectedLength: Int) {
```
Use CsvSource instead of MethodSource, to keep the test data local to the test.

Commit:
Changes for String Pattern throwOnFailure()

Commit:
Changes for String Pattern Csv Source Tests

Commit:
Removed unwanted arguments

Commit:
String Pattern Validation Updated

Commit:
Specmatic String Pattern Tests

Commit:
Specmatic String Pattern Tests Rectified

Commit:
Specmatic String Pattern Tests Rectified

Commit:
Merged latest

Commit:
Merged latest

Commit:
Merged latest

harikrishnan83 committed:
Merge branch 'main' into pattern_max_length_fix

* main: (124 commits)
  Fix schema regex and generation
  2.0.43
  Generate 4xx responses as per the schema in virtual-service be it any pattern containing a json object pattern
  Ignore a seed data entry which involves attribute selection during the seed data load stage of the virtual service
  Ignore a seed data entry with an id which is pre-existing in the cache of the virtual service
  Fix failing tests due to stale config.
  Add test for ExamplePreProcessor
  Add tests for Asserts.
  Cleanup on amber validation.
  2.0.42
  Changes for GoBack
  Added Details Pre Div
  Removed highlighting of specmatic errors in editor
  Editor syntax highlighting
  Fix full pattern validation with ListPattern.
  Add tests for partially valid examples.
  Better mismatchMessages for missing optional keys.
  better variable namings and partial handling
  Interactive Server partially valid examples.
  Modify result logic when all keys are mandatory.
  ...

harikrishnan83 committed:
- minor grammatical and spelling changes in tests
- adding tests for contradicting length constraints and regex

joelrosario committed:
Refactored StringPattern.matches

joelrosario committed:
Merge branch 'main' into pattern_max_length_fix
